### ISSUE TITLE
Dafny 1.9.8 not working on Ubuntu: "Fatal Unhandled Exception"

### ISSUE THREAD

**Issue by pablo-m-s-farias** [Microsoft/dafny#94]
Hi, all!

I am not succeeding at putting Dafny to work in my computer's Ubuntu 16.04 32 bits (with Mono version 4.8.0).

Today, having read the installation instructions for [using the binaries](https://github.com/Microsoft/dafny/wiki/INSTALL) and for [using the source](https://github.com/Microsoft/dafny/blob/master/INSTALL), I have installed Dafny in both ways. In the installation from the binaries, if I run the [dafny] script, I get

> /usr/bin/env: “bash\r”: File or folder not found

However, if I run the [dafny] script in the installation from the source, or if I run [mono Dafny.exe] in either of them, I get

> *** Error: No input files were specified.

which looks promising. But then if I save this code

`method test () { assert 10 < 20; }`

in file [test_assert.dfy] and run [mono Dafny.exe test_assert.dfy] I get

> Dafny program verifier version 1.9.8.30829, Copyright (c) 2003-2016, Microsoft.
> [ERROR] FATAL UNHANDLED EXCEPTION: System.AggregateException: One or more errors occurred. ---> System.IO.IOException: Write fault on path /home/pablo/Documentos/dafny_1-9-8/dafny/[Unknown]
>   at System.IO.FileStream.WriteInternal (System.Byte[] src, System.Int32 offset, System.Int32 count) [0x00086] in <f712f98eb8e445c8918edaf595bbe465>:0 
>   at System.IO.FileStream.Write (System.Byte[] array, System.Int32 offset, System.Int32 count) [0x000a5] in <f712f98eb8e445c8918edaf595bbe465>:0 
>   at System.IO.StreamWriter.Flush (System.Boolean flushStream, System.Boolean flushEncoder) [0x00094] in <f712f98eb8e445c8918edaf595bbe465>:0 
>   at System.IO.StreamWriter.Write (System.Char[] buffer, System.Int32 index, System.Int32 count) [0x000ee] in <f712f98eb8e445c8918edaf595bbe465>:0 
>   at System.IO.TextWriter.WriteLine (System.String value) [0x00083] in <f712f98eb8e445c8918edaf595bbe465>:0 
>   at Microsoft.Boogie.SMTLib.SMTLibProcess.Send (System.String cmd) [0x00073] in <0abaf91c118f4d46ac1a6a1d87774c56>:0 
>   at Microsoft.Boogie.SMTLib.SMTLibProcessTheoremProver.Send (System.String s, System.Boolean isCommon) [0x00036] in <0abaf91c118f4d46ac1a6a1d87774c56>:0 
>   at Microsoft.Boogie.SMTLib.SMTLibProcessTheoremProver.SendThisVC (System.String s) [0x00001] in <0abaf91c118f4d46ac1a6a1d87774c56>:0 
>   at Microsoft.Boogie.SMTLib.SMTLibProcessTheoremProver.Reset (Microsoft.Boogie.VCExpressionGenerator gen) [0x0001e] in <0abaf91c118f4d46ac1a6a1d87774c56>:0 
>   at Microsoft.Boogie.Checker.BeginCheck (System.String descriptiveName, Microsoft.Boogie.VCExprAST.VCExpr vc, Microsoft.Boogie.ProverInterface+ErrorHandler handler) [0x00025] in <b1b95669937649e3967d891cab555946>:0 
>   at VC.VCGen+Split.BeginCheck (Microsoft.Boogie.Checker checker, Microsoft.Boogie.VerifierCallback callback, VC.ModelViewInfo mvInfo, System.Int32 no, System.Int32 timeout) [0x0022b] in <b1b95669937649e3967d891cab555946>:0 
>   at VC.VCGen.VerifyImplementation (Microsoft.Boogie.Implementation impl, Microsoft.Boogie.VerifierCallback callback) [0x00716] in <b1b95669937649e3967d891cab555946>:0 
>    --- End of inner exception stack trace ---
>   at System.AggregateException.Handle (System.Func`2[T,TResult] predicate) [0x00076] in <f712f98eb8e445c8918edaf595bbe465>:0 
>   at Microsoft.Boogie.ExecutionEngine.InferAndVerify (Microsoft.Boogie.Program program, Microsoft.Boogie.PipelineStatistics stats, System.String programId, Microsoft.Boogie.ErrorReporterDelegate er, System.String requestId) [0x004d4] in <ed4f363be8a84ed0b60d9c6cb73d3832>:0 
>   at Microsoft.Dafny.DafnyDriver.BoogiePipelineWithRerun (Microsoft.Boogie.Program program, System.String bplFileName, Microsoft.Boogie.PipelineStatistics& stats, System.String programId) [0x0010e] in <65445563cb7e4b178d388e1e97a4da77>:0 
>   at Microsoft.Dafny.DafnyDriver.Boogie (System.Collections.Generic.IList`1[T] dafnyFileNames, Microsoft.Boogie.Program boogieProgram, System.String programId, Microsoft.Boogie.PipelineStatistics& stats, Microsoft.Boogie.PipelineOutcome& oc) [0x00095] in <65445563cb7e4b178d388e1e97a4da77>:0 
>   at Microsoft.Dafny.DafnyDriver.ProcessFiles (System.Collections.Generic.IList`1[T] dafnyFileNames, System.Collections.ObjectModel.ReadOnlyCollection`1[T] otherFileNames, Microsoft.Dafny.ErrorReporter reporter, System.Boolean lookForSnapshots, System.String programId) [0x0029c] in <65445563cb7e4b178d388e1e97a4da77>:0 
>   at Microsoft.Dafny.DafnyDriver.ThreadMain (System.String[] args) [0x0007b] in <65445563cb7e4b178d388e1e97a4da77>:0 
>   at Microsoft.Dafny.DafnyDriver+<>c__DisplayClass1.<Main>b__0 () [0x00001] in <65445563cb7e4b178d388e1e97a4da77>:0 
>   at System.Threading.ThreadHelper.ThreadStart_Context (System.Object state) [0x00017] in <f712f98eb8e445c8918edaf595bbe465>:0 
>   at System.Threading.ExecutionContext.RunInternal (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) [0x0008d] in <f712f98eb8e445c8918edaf595bbe465>:0 
>   at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) [0x00000] in <f712f98eb8e445c8918edaf595bbe465>:0 
>   at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state) [0x00031] in <f712f98eb8e445c8918edaf595bbe465>:0 
>   at System.Threading.ThreadHelper.ThreadStart () [0x0000b] in <f712f98eb8e445c8918edaf595bbe465>:0 
> ---> (Inner Exception #0) System.IO.IOException: Write fault on path /home/pablo/Documentos/dafny_1-9-8/dafny/[Unknown]
>   at System.IO.FileStream.WriteInternal (System.Byte[] src, System.Int32 offset, System.Int32 count) [0x00086] in <f712f98eb8e445c8918edaf595bbe465>:0 
>   at System.IO.FileStream.Write (System.Byte[] array, System.Int32 offset, System.Int32 count) [0x000a5] in <f712f98eb8e445c8918edaf595bbe465>:0 
>   at System.IO.StreamWriter.Flush (System.Boolean flushStream, System.Boolean flushEncoder) [0x00094] in <f712f98eb8e445c8918edaf595bbe465>:0 
>   at System.IO.StreamWriter.Write (System.Char[] buffer, System.Int32 index, System.Int32 count) [0x000ee] in <f712f98eb8e445c8918edaf595bbe465>:0 
>   at System.IO.TextWriter.WriteLine (System.String value) [0x00083] in <f712f98eb8e445c8918edaf595bbe465>:0 
>   at Microsoft.Boogie.SMTLib.SMTLibProcess.Send (System.String cmd) [0x00073] in <0abaf91c118f4d46ac1a6a1d87774c56>:0 
>   at Microsoft.Boogie.SMTLib.SMTLibProcessTheoremProver.Send (System.String s, System.Boolean isCommon) [0x00036] in <0abaf91c118f4d46ac1a6a1d87774c56>:0 
>   at Microsoft.Boogie.SMTLib.SMTLibProcessTheoremProver.SendThisVC (System.String s) [0x00001] in <0abaf91c118f4d46ac1a6a1d87774c56>:0 
>   at Microsoft.Boogie.SMTLib.SMTLibProcessTheoremProver.Reset (Microsoft.Boogie.VCExpressionGenerator gen) [0x0001e] in <0abaf91c118f4d46ac1a6a1d87774c56>:0 
>   at Microsoft.Boogie.Checker.BeginCheck (System.String descriptiveName, Microsoft.Boogie.VCExprAST.VCExpr vc, Microsoft.Boogie.ProverInterface+ErrorHandler handler) [0x00025] in <b1b95669937649e3967d891cab555946>:0 
>   at VC.VCGen+Split.BeginCheck (Microsoft.Boogie.Checker checker, Microsoft.Boogie.VerifierCallback callback, VC.ModelViewInfo mvInfo, System.Int32 no, System.Int32 timeout) [0x0022b] in <b1b95669937649e3967d891cab555946>:0 
>   at VC.VCGen.VerifyImplementation (Microsoft.Boogie.Implementation impl, Microsoft.Boogie.VerifierCallback callback) [0x00716] in <b1b95669937649e3967d891cab555946>:0 <---
> 
> ---> (Inner Exception #1) System.IO.IOException: Write fault on path /home/pablo/Documentos/dafny_1-9-8/dafny/[Unknown]
>   at System.IO.FileStream.WriteInternal (System.Byte[] src, System.Int32 offset, System.Int32 count) [0x00086] in <f712f98eb8e445c8918edaf595bbe465>:0 
>   at System.IO.FileStream.Write (System.Byte[] array, System.Int32 offset, System.Int32 count) [0x000a5] in <f712f98eb8e445c8918edaf595bbe465>:0 
>   at System.IO.StreamWriter.Flush (System.Boolean flushStream, System.Boolean flushEncoder) [0x00094] in <f712f98eb8e445c8918edaf595bbe465>:0 
>   at System.IO.StreamWriter.Write (System.Char[] buffer, System.Int32 index, System.Int32 count) [0x000ee] in <f712f98eb8e445c8918edaf595bbe465>:0 
>   at System.IO.TextWriter.WriteLine (System.String value) [0x00083] in <f712f98eb8e445c8918edaf595bbe465>:0 
>   at Microsoft.Boogie.SMTLib.SMTLibProcess.Send (System.String cmd) [0x00073] in <0abaf91c118f4d46ac1a6a1d87774c56>:0 
>   at Microsoft.Boogie.SMTLib.SMTLibProcessTheoremProver.Send (System.String s, System.Boolean isCommon) [0x00036] in <0abaf91c118f4d46ac1a6a1d87774c56>:0 
>   at Microsoft.Boogie.SMTLib.SMTLibProcessTheoremProver.SendThisVC (System.String s) [0x00001] in <0abaf91c118f4d46ac1a6a1d87774c56>:0 
>   at Microsoft.Boogie.SMTLib.SMTLibProcessTheoremProver.Reset (Microsoft.Boogie.VCExpressionGenerator gen) [0x0001e] in <0abaf91c118f4d46ac1a6a1d87774c56>:0 
>   at Microsoft.Boogie.Checker.BeginCheck (System.String descriptiveName, Microsoft.Boogie.VCExprAST.VCExpr vc, Microsoft.Boogie.ProverInterface+ErrorHandler handler) [0x00025] in <b1b95669937649e3967d891cab555946>:0 
>   at VC.VCGen+Split.BeginCheck (Microsoft.Boogie.Checker checker, Microsoft.Boogie.VerifierCallback callback, VC.ModelViewInfo mvInfo, System.Int32 no, System.Int32 timeout) [0x0022b] in <b1b95669937649e3967d891cab555946>:0 
>   at VC.VCGen.VerifyImplementation (Microsoft.Boogie.Implementation impl, Microsoft.Boogie.VerifierCallback callback) [0x00716] in <b1b95669937649e3967d891cab555946>:0 <---
> 

Any clue about what is happening?

**Comment by parno** [comment]
For the first error, you may need to run `dos2unix` on the dafny script to get rid of MS-style line endings.  The second error may be a result of not being able to invoke Z3 successfully.

**Comment by pablo-m-s-farias** [comment]
Thanks, @parno. Running `dos2unix` did solve the first error.

As for the second one, I don't know what to do. I did, however, try to run z3 independently of Dafny: executing file `dafny/Binaries/z3/bin/z3` results in

> bash: dafny/Binaries/z3/bin/z3: cannot execute binary file: Erro no formato exec

(The portuguese "Erro no formato exec" meaning "error in exec format".)

Note that this Z3 was installed from source as indicated in the instructions:

> cd BASE-DIRECTORY
> wget https://github.com/Z3Prover/z3/releases/download/z3-4.4.0/z3-4.4.0-x64-ubuntu-14.04.zip
> unzip z3-4.4.0-x64-ubuntu-14.04.zip
> mv z3-4.4.0-x64-ubuntu-14.04 dafny/Binaries/z3

I get the same result when trying to execute the z3 file that came with the Dafny 1.9.8 binaries.

**Comment by pablo-m-s-farias** [comment]
Sorry, I now realize that the procedure above for the Z3 installation is not "from source" as I said, but instead just copying the zip file from github.

**Comment by parno** [comment]
You might try asking the Z3 folks about this.  It doesn't sound like it's Dafny specific at this point.  Once you're able to run Z3 directly, Dafny (via Boogie) ought to be able to as well.

**Comment by pablo-m-s-farias** [comment]
I've just solved the problem: the Z3 error was happening simply because I mechanically repeated the installation instructions, thereby not realizing that I was downloading the 64 bits version of Z3, instead of the 32 bits version needed in my case. I finally managed to successfully run a local installation of Dafny.

I make the following suggestions:

  * In the [installation instructions](https://github.com/Microsoft/dafny/blob/master/INSTALL.md), warn the user to choose the 32 bits version of Z3 if this is the case.

  * Warn the user to run `chmod u+x Dafny.exe` if needed. Before I did that, when I ran the `dafny` shell script, the system reported: `Error: Dafny.exe not found at ./Dafny.exe.`.

Thanks for the help, @parno! Pointing out that the problem had to do with Z3 was very useful.
